### PR TITLE
WT-8792 Checkpoint cleanup never frees the on-disk pages with overflow items 

### PR DESCRIPTION
**Pull request by sauclovian-wt**
During checkpoint cleanup bring in a few obsolete pages with overflow items.

This causes them to get reconciled and discarded while still cleaning up the overflow items correctly.

Note: there's throttle logic here to keep from blowing out the cache by reading too much junk in, but it's arbitrary and untuned and possibly too aggressive.

### CODE REVIEW

**Comment by lukech** [comment]
autobuild: Can one of the admins verify this patch?

**Comment by sauclovian-wt** [comment]
I figured the throttle logic should be affected by the size of the tree (not just a fixed constant number of pages at a time) but I suspect I've made it too aggressive and feedback would be helpful.

**Comment by sauclovian-wt** [comment]
The new test is numbered on the assumption that the large number of new test_checkpoint* in WT-5927 are going to get merged eventually.

File: test/suite/test_checkpoint23.py
Comment (etienneptl) [inline test/suite/test_checkpoint23.py:141]: The content seems to be duplicated from here

Comment (sauclovian-wt) [reply test/suite/test_checkpoint23.py]: Wonder how that happened... fixed in bd5b940

File: src/btree/bt_sync.c
Comment (keithbostic) [inline src/btree/bt_sync.c:152]: Returns need parentheses.

File: src/btree/bt_sync.c
Comment (keithbostic) [inline src/btree/bt_sync.c:166]: Return needs parentheses.

File: src/btree/bt_sync.c
Comment (keithbostic) [inline src/btree/bt_sync.c:164]: I would just use `__wt_cache_stuck()` here, that's what compaction uses.

File: src/btree/bt_sync.c
Comment (keithbostic) [inline src/btree/bt_sync.c:254]: I'd recommend using `if (ret == 0) {`, I think that's more obvious with respect to the error handling.

File: src/btree/bt_sync.c
Comment (keithbostic) [inline src/btree/bt_sync.c:252]: I'm confused by `WT_READ_SKIP_DELETED`: can that actually be set on an overflow page, and does it matter?

File: src/btree/bt_sync.c
Comment (keithbostic) [inline src/btree/bt_sync.c:470]: `load_throttle` can just be a `u_int`, or even `int`

Comment (keithbostic) [inline src/btree/bt_sync.c:470]: Since `load_throttle` is overflow specific, maybe it needs a more overflow specific name?

File: src/btree/bt_sync.c
Comment (keithbostic) [inline src/btree/bt_sync.c:611]: The `__sync_load_throttle_init` call could be moved into `__sync_ref_int_obsolete_cleanup()`, and combined with the `__sync_load_throttle_next_int` call as far as I can tell, and there's no reason we couldn't just do the initialization inline there?

This seems more complicated that it needs to be: `__sync_should_read_obsolete_page()` can be combined with `__sync_load_throttle_check()`, too.

IOW, I think we only need a single helper function.

**Review by keithbostic — APPROVED** [review]
OK with me.

I don't feel strongly about this function -- now that I understand why it's here, I'm not eager to merge the work, but on the other hand, the code is already written and it's not particularly intrusive, so I'm good to go.

File: test/suite/test_checkpoint23.py
Comment (kommiharibabu) [inline test/suite/test_checkpoint23.py:118]: I don't clearly understand the need for the second reopen_conn(), Is the recovery loads the overflow pages back into memory? Do we really need that?

Comment (sauclovian-wt) [reply test/suite/test_checkpoint23.py:118]: The stat pass will load the tree into memory to count the overflow items. (I believe. At least, the code in bt_stat.c that does the counting uses in-memory pages.)

File: src/btree/bt_sync.c
Comment (sauclovian-wt) [reply src/btree/bt_sync.c]: Fixed in 2fbb617ec (and obsoleted in 395664db3)

File: src/btree/bt_sync.c
Comment (sauclovian-wt) [reply src/btree/bt_sync.c]: Fixed in 2fbb617ec (and obsoleted in 395664db3)

File: src/btree/bt_sync.c
Comment (sauclovian-wt) [reply src/btree/bt_sync.c:126]: (After discussion offline we decided to keep this set of predicates because it's the same as used in __sync_page_skip below.)

File: src/btree/bt_sync.c
Comment (sauclovian-wt) [reply src/btree/bt_sync.c]: Fixed in 2fbb617ec.

File: src/btree/bt_sync.c
Comment (sauclovian-wt) [reply src/btree/bt_sync.c:217]: Since we unlock the ref, in theory the page could be deleted by someone else before we get it back. Not likely, but if so there's no point reinstantiating it. Added a comment in 12e278cd9.

File: src/btree/bt_sync.c
Comment (sauclovian-wt) [reply src/btree/bt_sync.c]: Fixed in 2fbb617ec (and obsoleted in 395664db3)

Comment (sauclovian-wt) [reply src/btree/bt_sync.c]: Fixed in 2fbb617ec (now overflow_load_throttle) and obsoleted in 395664db3.

File: src/btree/bt_sync.c
Comment (sauclovian-wt) [reply src/btree/bt_sync.c]: The goal was to have the entire throttle computation in one place, because it's currently arbitrary stuff I made up and so it's reasonably likely that someone sometime will want to rearrange it. Also, it turns out the init call was in the wrong place; it was supposed to be before the treewalk starts, not once per page, so those two pieces aren't actually the same. (Fixed this in 7b54c135e.) (Whole thing obsoleted in 395664db3.)

File: test/suite/test_checkpoint23.py
Comment (sauclovian-wt) [reply test/suite/test_checkpoint23.py:119]: (I checked; it doesn't work without the second reopen. Added a comment in b27fd50b0.)

Comment (kommiharibabu) [reply test/suite/test_checkpoint23.py:119]: Thanks for confirming it.

**Review by kommiharibabu — APPROVED** [review]
LGTM

**Comment by sauclovian-wt** [comment]
After discussion with Keith we decided to remove the explicit throttle logic (the counting stuff), keeping just the checks that eviction is happy; as long as that stays true, chug away.

The rationale is: (a) this situation probably isn't going to come up much in practice so it isn't likely to hit very often or on very many pages at a time; (b) consequently we don't have any reasonable test workloads for guidance regarding the throttle logic; (c) the throttle logic I made up was just something I arbitrarily made up and there's no particular reason to think it's good or bad or different from any other arbitrary choice; and (d) in the absence of any compelling argument to the contrary keeping the code simple is as good a rationale as anything else.

I did this as a separate commit (and fixed all the review comments so far in the affected code first) so if consensus is otherwise it can be reverted easily.

But, take note; if you object, speak up...

File: test/suite/test_checkpoint23.py
Comment (sauclovian-wt) [reply test/suite/test_checkpoint23.py:119]: YW :-)

File: src/btree/bt_sync.c
Comment (agorrod) [inline src/btree/bt_sync.c:211]: I was all prepared to come back and comment on the throttle logic, then found that it was gone. That decision is fine with me. I'm happy to optimize if there is a need in the future.

Could you update this comment?

**Review by agorrod — APPROVED** [review]
I'm happy with this change modulo a request for a change to a comment. Nice catch, and well done figuring out the path - I always find overflow pages on disk a bit of a head scratcher. Also: thanks for adding a stat!
Don't wait for another round of review from me.

**Comment by agorrod** [comment]
@sauclovian-wt is this waiting on anything before merge? (hopefully not me, but if it was LGTM)

**Comment by sauclovian-wt** [comment]
(I think it was just nobody got to it Friday evening)

**Pull request merged**
